rfcs/0031-bazel.md

```diff
@@ -0,0 +1,358 @@
+      maintainers to accept the Bazel-enabling patches.
+    * For Coda this approach may or may not be desirable.  The
+      alternative is to use the same standard approach as used by the
+      other external libs, namely an external BUILD file.
+  * Internal C/C++ code.  The Code codebase contains some OCaml code
+    that generates C/C++ code which it then compiles.  The task is to
+    convert this to C/C++ code and let Bazel handle any customizations
+    needed.
+  * CUDA.  Coda contains a little bit of CUDA code.  I assume Bazel
+    will easily handle this, but I do not have a GPU, and have never
+    worked with CUDA.
+  * Cross-compile toolchains.  I've got a minimal example working
+    using a crosstool-NG toolchain on a Mac. The task is to polish it
+    and port it to Linux and possibly Windows.
+* OCaml
+  * Learn enough OCaml/Dune to understand the OCaml/C++ interface source and build code
+  * Learn enough Dune to understand how it is used in the Coda codebase.
+  * Analyze existing Bazel libraries supporting OCaml
+    * https://github.com/jin/rules_ocaml
+    * https://github.com/ostera/rules_reason
+  * develop `rules_ocaml`
+    * Option 1: Integrate Bazel and Dune
+      * Analyze `rules_foreign_cc` to see how it handles integration of
+    configure and make.
+        * Prototype a few Minimal Working Examples of using Bazel together with Dune.
+        * Implement `rules_ocaml` using this integrational strategy
+    * Option 2: Bazel-only ruleset
+      * Analyze `rules_go` (Bazel rules for Golang) as a possible model for `rules_ocaml`
+      * Implement `rules_ocaml` using this integrational strategy
+  * Use the new `rules_ocaml` to implement bazel support for Coda codebase:
+    * External libs
+    * Internal OCaml code
+    * C/C++ interop code
+* Golang
+  * Coda uses the Golang implementation of libp2p. Google maintains an
+    official set of Bazel rules for Go ([rules_go](https://github.com/bazelbuild/rules_go), so this should be
+    straightforward.
+
+**STATUS**
+
+* All of the external C/C++ libs have been bazelized, except one,
+  Postgresql, which actually builds correctly, but then fails in the
+  Bazel code.  I haven't figured out a fix yet; the workaround is to
+  just follow current practice and install Postgresql separately.
+* I've got some cross-compile code from a previous project that
+  supports Mac->Linux (using a crosstools-NG toolchain) and
+  Mac->Android (using the Android NDK toolchain).  Bazel has changed
+  its toolchain model, so these must be converted. I've done that with
+  the crosstools-NG toolchain and managed a minimal build of some of
+  the libraries with it.  It will require a little more work and
+  testing, but basically we're in good shape as far as C/C++ cross
+  compilation is concerned.
+* Haven't really looked into Bazel-Ocaml yet.
+
+### Effort Estimates and Timeline
+
+Effort = estimated remaining work effort in hours.
+
+C/C++
```

Comment:
Cross builds to Linux are great if it's easy enough. ARM and Android cross definitely not required (plus you'll have a _really_ hard time getting it to work with a newer version of the OCaml compiler)